bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/device/TapoDevice.java

```diff
@@ -0,0 +1,429 @@
+import org.openhab.binding.tapocontrol.internal.helpers.TapoCredentials;
+import org.openhab.binding.tapocontrol.internal.helpers.TapoErrorHandler;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract class as base for TAPO-Device device implementations.
+ *
+ * @author Christian Wild - Initial contribution
+ */
+@NonNullByDefault
+public abstract class TapoDevice extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(TapoDevice.class);
+    private final HttpClient httpClient;
+    protected final String uid;
+    private String ipAddress = "";
+    // protected @NonNullByDefault({}) TapoDeviceConnector connector;
+    protected @NonNullByDefault({}) TapoControlConfiguration config;
+    protected @Nullable ScheduledFuture<?> pollingJob;
+    protected @Nullable TapoDeviceConnector connector;
+    protected TapoCredentials credentials;
+
+    /**
+     * Constructor
+     *
+     * @param thing Thing object representing device
+     */
+    public TapoDevice(Thing thing, HttpClient httpClient) {
+        super(thing);
+        this.uid = getThing().getUID().getAsString();
+        this.credentials = new TapoCredentials();
+        this.httpClient = httpClient;
+    }
+
+    /***********************************
+     *
+     * INIT AND SETTINGS
+     *
+     ************************************/
+
+    /**
+     * INITIALIZE DEVICE
+     */
+    @Override
+    public void initialize() {
+        logger.debug("initialize thing-device ({})", uid);
+        String username = "";
+        String password = "";
+        try {
+            Bridge bridge = this.getBridge();
+            username = bridge.getConfiguration().get(CONFIG_EMAIL).toString();
+            password = bridge.getConfiguration().get(CONFIG_PASS).toString();
+            this.credentials.setCredectials(username, password);
+            this.ipAddress = getThing().getConfiguration().get(CONFIG_DEVICE_IP).toString();
+        } catch (Exception e) {
+            logger.debug("({}) configuration error : {}", uid, e.getMessage());
+        }
+        TapoErrorHandler configError = checkSettings();
+        if (!configError.hasError()) {
+            this.connector = new TapoDeviceConnector(this, credentials, httpClient);
+
+            // set the thing status to UNKNOWN temporarily and let the background task decide for the real status.
+            updateStatus(ThingStatus.UNKNOWN);
+
+            // background initialization (delay it a little bit):
+            scheduler.schedule(this::connect, 2000, TimeUnit.MILLISECONDS);
```

Comment:
This task needs to be cancelled in case the handler is started and stopped again immediately.